rotkehlchen/exchanges/okx.py

```diff
@@ -0,0 +1,410 @@
+from urllib.parse import urlencode, urljoin
+
+import requests
+
+from rotkehlchen.accounting.ledger_actions import LedgerAction
+from rotkehlchen.accounting.structures.balance import Balance
+from rotkehlchen.assets.asset import AssetWithOracles
+from rotkehlchen.assets.converters import asset_from_okx
+from rotkehlchen.constants import ZERO
+from rotkehlchen.errors.asset import UnknownAsset, UnsupportedAsset
+from rotkehlchen.errors.misc import RemoteError
+from rotkehlchen.errors.serialization import DeserializationError
+from rotkehlchen.exchanges.data_structures import AssetMovement, MarginPosition, Trade
+from rotkehlchen.exchanges.exchange import ExchangeInterface, ExchangeQueryBalances
+from rotkehlchen.exchanges.utils import deserialize_asset_movement_address
+from rotkehlchen.history.deserialization import deserialize_price
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.serialization.deserialize import deserialize_asset_amount, deserialize_fee
+from rotkehlchen.types import (
+    ApiKey,
+    ApiSecret,
+    AssetMovementCategory,
+    Fee,
+    Location,
+    Timestamp,
+    TradeType,
+)
+from rotkehlchen.user_messages import MessagesAggregator
+
+if TYPE_CHECKING:
+    from rotkehlchen.db.dbhandler import DBHandler
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class Okx(ExchangeInterface):
+    """
+    OKX exchange API docs:
+    https://www.okx.com/docs-v5
+    """
+
+    MAX_RESULTS = 100
+
+    def __init__(
+            self,
+            name: str,
+            api_key: ApiKey,
+            secret: ApiSecret,
+            passphrase: str,
+            database: 'DBHandler',
+            msg_aggregator: MessagesAggregator,
+    ):
+        super().__init__(
+            name=name,
+            location=Location.OKX,
+            api_key=api_key,
+            secret=secret,
+            database=database,
+        )
+        self.passphrase = passphrase
+        self.base_uri = 'https://www.okx.com/'
+        self.msg_aggregator = msg_aggregator
+        self.session.headers.update({
+            'OK-ACCESS-KEY': self.api_key,
+            'OK-ACCESS-PASSPHRASE': self.passphrase,
+        })
+
+    def edit_exchange_credentials(
+            self,
+            api_key: Optional[ApiKey],
+            api_secret: Optional[ApiSecret],
+            passphrase: Optional[str],
+    ) -> bool:
+        changed = super().edit_exchange_credentials(api_key, api_secret, passphrase)
+        if api_key is not None:
+            self.session.headers.update({'OK-ACCESS-KEY': self.api_key})
+        if passphrase is not None:
+            self.session.headers.update({'OK-ACCESS-PASSPHRASE': self.passphrase})
+        return changed
+
+    def _generate_signature(self, timestamp: str, method: str, path: str, body: str) -> str:
+        prehash = timestamp + method.upper() + path + body
```

Comment:
can you link the url where it explains how to build this one? for future references